File: leet_1461_contain_all_bin_comb___set.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        st = set()
        for i in range(len(s) + 1 - k):
            st.add(s[i : i + k])
        return len(st) == 2 ** k

# same code, above is faster

"""
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        n = len(s)
        if n < k:
            return False
        st = set()
        for i in range(n - k + 1):
            st.add(s[i : i + k])
        if len(st) == 2 ** k:
            return True
        return False
"""

File: leet_1461_contain_all_bin_comb___set.cpp
class Solution {
public:
    bool hasAllCodes(string s, int k)
    {
        int         n = s.length();

        if (n < k)
        {
            return false;
        }

        set<string> st;
        int         i = -1;

        while (++i < n - k + 1)
        {
            st.insert(s.substr(i, k));
        }
        if (st.size() == pow(2, k))
        {
            return true;
        }
        return false;
    }
};
